Move profiler to a dispatch wrapper

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33057 Move profiler to a dispatch wrapper**

This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

Commit:
[WIP] Move profiler to a dispatch wrapper

[ghstack-poisoned]

ilia-cher commented:
cool, looks interesting, could you add a bit more on the motivation, is it to expand coverage of ops that can be profiled? I'm also curious if there's any change in overhead

smessmer commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -340,3 +347,45 @@ void RecordProfile::processEvents(const std::vector<Event*>& events) {
+  // Instead, we take a path through the JIT operator registry, which has a boxed
+  // calling mechanism that works for all ops from native_functions.yaml.
+
+  auto s = c10::Symbol::fromQualString(op.schema().name());
```
warning: this re-dispatch is probably slow. But it's the only way that works for 100% of ops currently.

ezyang replied:
At the very least, we can speed up the process of looking up the JIT operator, if we introduce some sort of mapping from OperatorHandle to JIT schema. @smessmer do you know if this can be conveniently done?

ezyang replied:
@smessmer  We need to make a call on what should be happening here. @jamesr66a do you think you could characterize when the boxed workaround gets called? It's going to just be for custom operators which didn't have VariableType generated for them. How were these supported before?

I just really don't think it's acceptable to land a silent slowdown that will affect all custom operators when you're benchmarking. It would be better to just not have accurate profiling information in this case at all.

smessmer replied:
we are relatively close to making the non-workaround callBoxed work for redispatching, but it'll likely take a few more days

jamesr66a replied:
How about for now, I remove this and reinstate the `RECORD_FUNCTION` at https://github.com/pytorch/pytorch/pull/33057/files#diff-94205457152f6821c30329a4ddb40bbeL21, and then we can add this back with the proper callBoxed call later?

smessmer commented on c10/core/DispatchKey.h:
```diff
@@ -59,6 +59,8 @@ enum class DispatchKey : uint8_t {
 
   VariableTensorId,
 
+  ProfilerId,
```
Do we want this after autograd? Afaik, the old profiling was before autograd

jamesr66a replied:
I'm a bit confused about the ordering here. I was under the impression that this is a higher bit index since it's later than the enum. Above it says higher bit indexes are dispatched first because of the leading zeros thing

ezyang replied:
This is the correct position. Lower in the enum = happens earlier / higher priority

ezyang replied:
We could, although it would impact the code we generate (as we now have to bitshift in the other direction)

smessmer reviewed (approved):
> In general, does this look like a reasonable approach?

Yes, this looks pretty good to me

> There seems to be the concept of "unboxed only" kernels, would those work with this system?

Yes, this should work with them. Making them work is the whole reason we do this `callBoxedWorkaround` instead of just calling `op.callBoxed()`.

> (probably a broader design question) Can we further extend this to include stuff like prim:: ops? (which currently are sometimes/piecemeal supported in profiling)

Afaik, the plan for prim ops is to make some into real ops, then this would work for them too, and make the rest into actual interpreter byte code instructions, for those this would not work.

ezyang commented:
Thank you for working on this. Hopefully, this approach will be reasonable to do in the long term, but in the short term there are problems which prevent this from directly going in:

1. Not all operators support boxing/unboxing wrappers. @smessmer can say better about what is and isn't supported, but I believe that as of today, an implementation done this way will most prominently lose profiling on all inplace operations (e.g., `x.add_`). @mcarilli also has relevant experience here, as he was using the boxed fallback mechanism previously to implement fallthrough, and he saw his operator support shoot up when I implemented native support for boxed fallback.  According to our roadmap, we are committed to 100% templated unboxing by the end of this half, but I don't know what our target is for *boxing*, which is what would be needed here. Things are less bad if you only need to support profiling from the JIT, since you don't need to box in that case (everything is already boxed)
2. There are big performance problems with the patch. The most flagrant one is the workaround for callBoxed not working, but this should be solved at the end of this half as smessmer says. A more subtle performance problem is for all codepaths that come into the dispatcher from the unboxed calling interface (that's all of them, but JIT moving over to the boxed path by the end of this half), you have to *box* all the inputs, run the boxed fallback, and unbox them again. And of course, there's the cost of doing an extra dispatch to get to this kernel (40ns, according to @ngimel)

My recommendation when I've discussed this with @ngimel is that we should continue to codegen unboxed versions of the `RECORD_FUNCTION` calls in the short term (it may be possible to metaprogram them).  Otherwise you have to wait for a big pile of infra work before you can land this.

ezyang reviewed (changes requested):
I think the accept is a little premature :>

jamesr66a commented:
@ilia-cher yeah the motivation is to expand coverage as well as make handling of profiling go through fewer codepaths (e.g. the commentary about moving `prim::` ops to this path

ezyang commented:
We could also hard code profiling into the dispatcher in the same way we hard code fallthrough, **IF** you don't care about passing arguments to `RECORD_FUNCTION`. But I bet the tensor arguments are pretty useful...

jamesr66a commented:
@ezyang yeah having the arguments is pretty critical for some other functionality (e.g. dataflow analysis) we might implement so baking it into the dispatcher might shoot us in the foot in that regard. I'm going to think through the codegen proposal, might be a good meantime solution

jamesr66a commented:
And as a point of information, I actually just found that this patch as-is crashes on in-place ops:

```
import torch

x = torch.rand(3, 4)

with torch.autograd.profiler.profile() as prof:
    torch.relu_(x)
print(prof)
```

```
Traceback (most recent call last):
  File "/home/jamesreed/relu.py", line 6, in <module>
    torch.relu_(x)
RuntimeError: false INTERNAL ASSERT FAILED at ../aten/src/ATen/core/boxing/boxing.h:53, please report a bug to PyTorch. Tried to call KernelFunction::callUnboxed() for a kernel that only has a boxed kernel and doesn't support calling from an unboxed API yet.
```

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

jamesr66a commented:
Just updated to include the codegen'ed wrappers. I also disabled the fallback wrapper because of the issue mentioned in https://github.com/pytorch/pytorch/pull/33057#issuecomment-583136505. I'll keep plugging away at this tomorrow morning

ezyang commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -340,3 +347,45 @@ void RecordProfile::processEvents(const std::vector<Event*>& events) {
+// auto registry = c10::Dispatcher::singleton()
+//   .registerBackendFallbackKernel(
+//     c10::DispatchKey::ProfilerId,
+//     c10::KernelFunction::makeFromBoxedFunction<&profile_wrapper>()
```
I actually agree with @smessmer's earlier comment, which is that we can keep the backend fallback to implement profiling for user-defined operators that don't have codegenned wrappers. It doesn't interfere with more specific unboxed implementation you are codegenning

ezyang commented on c10/core/DispatchKey.h:
```diff
@@ -59,6 +59,8 @@ enum class DispatchKey : uint8_t {
 
   VariableTensorId,
 
+  ProfilerId,
```
This is the nittiest of nits but I think the old convention of appending "Id" on every dispatch key doesn't make sense anymore because it's not TensorTypeId anymore, it's DispatchKey, so we should just drop the suffix and call it straight "DispatchKey::Profiler"

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -537,16 +574,82 @@ def gen_variable_type_shard(out, aten_declarations, template_path, suffix, heade
                     declaration, formal_types=formal_types))
 
+            # TODO: ????? type mismatch in generated wrapper?
+            if declaration['name'] != 'polygamma':
```
Weeeeird

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -537,16 +574,82 @@ def gen_variable_type_shard(out, aten_declarations, template_path, suffix, heade
+    for a in arguments:
+        processed_args.append('{}'.format(a['name']))
+
+    if is_tensor_method(declaration):
```
Not actionable comment on this PR: @smessmer, we should probably have some way of uniformly redispatching that doesn't require you to know if dispatch is going through a method or namespace.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -311,6 +312,39 @@
+{
+    AutoNonProfileTypeMode non_prof_type_mode;
+    RECORD_FUNCTION("${name}", std::vector<c10::IValue>({${input_names}}), Node::peek_at_next_sequence_nr());
+    return ${base_type_call};
```
I think it's OK to unconditionally use this template, as using a return in a void function with an expression is fine if the expression itself returns void. Citation: https://en.cppreference.com/w/cpp/language/return

ezyang reviewed (commented):
Good. I'd suggest:

1. Either deleting the backend fallback entirely, or turning it back on and cleaning up callBoxedWorkaround to just do callBoxed. Ideally we'd add a test to see that this is working for user defined functions.
2. A little performance benchmark would be great. Unlike before, this PR adds (1) a new dispatch and (2) a new pair of TLS toggles, and I want to know how much dispatch is costing us now. I think even if it's 40ns we should still land this patch, but it would put us in a good position to figure out how much more dispatcher optimization we need to do (there's is certainly more juice to squeeze from this stone).

jamesr66a commented:
Interestingly, with my changes, I'm seeing many new operators show up in the profile, and particularly not operators that were listed in DONT_PROFILE here https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_type.py#L89

For example TestAutograd.test_record_function expects:
```
            'profiler::_record_function_enter',
            'outer',
            'mul',
            'add',
            'profiler::_record_function_enter',
            'inner',
            'sub',
            'profiler::_record_function_exit',
            'profiler::_record_function_exit',
            'div',
```

but we get

```
profiler::_record_function_enter      
outer                                 
empty                                 
is_complex                            
mul                                   
is_complex                            
add                                   
profiler::_record_function_enter      
inner                                 
empty                                 
is_complex                            
sub                                   
profiler::_record_function_exit       
profiler::_record_function_exit       
is_complex                            
div           
```       

======

From offline meeting discussion: we've determined:

* This is due to things like `empty` not going through the VariableType wrappers, but now that profiling is done separately it shows up
* On the whole this is fine
* We'll have to deal with this in the future when we separate out tracing. A potential solution might be to make AutoNonVariableTypeMode also disable the tracing dispatch key (or whatever it ends up being)

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -179,7 +179,7 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 at::AutoNonVariableTypeMode _var_guard(true);
 switch(dispatchKeyToBackend(c10::impl::dispatchTypeId(${key_set},
-                            c10::DispatchKeySet(c10::DispatchKeySet::FULL).remove(DispatchKey::BackendSelect)))) {
+                            c10::DispatchKeySet(c10::DispatchKeySet::FULL).remove(DispatchKey::BackendSelect).remove(DispatchKey::Profiler)))) {
```
@ezyang I added this because it sounded like what you were referring to on the VC, but it looks like the mobile tests are still failing in `dispatchKeyToBackend`. I guess I'll try to get to debugging this

ezyang replied:
Yeah, removal here is what I would have expected would work. Odd.

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Questions:

1. In general, does this look like a reasonable approach?
2. There seems to be the concept of "unboxed only" kernels, would those work with this system?
3. (probably a broader design question) Can we further extend this to include stuff like `prim::` ops? (which currently are sometimes/piecemeal supported in profiling)

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "[WIP] Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

ezyang commented on c10/core/DispatchKey.cpp:
```diff
@@ -44,6 +44,8 @@ const char* toString(DispatchKey t) {
     case DispatchKey::TESTING_ONLY_GenericWrapperTensorId:
       return "TESTING_ONLY_GenericWrapperTensorId";
+    case DispatchKey::Profiler:
+      return "ProfileId";
```
Just return Profiler here. Also nb: https://github.com/pytorch/pytorch/pull/36240 is coming

ezyang commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -856,6 +856,10 @@ void testRecordFunction() {
 
+  bool orig_profiler =
+      c10::impl::tls_is_dispatch_key_included(c10::DispatchKey::Profiler);
+  c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Profiler, true);
```
Hmm, couldn't the RAII guard have been used here?

ezyang commented on test/test_autograd.py:
```diff
@@ -2792,14 +2808,38 @@ def forward(x):
+            'to',
+            'empty_strided',
+            'copy_',
+            'empty'
```
We should seriously consider putting this in an expect test; now that we are profiling a lot more, it seems likely that this will wobble as people make small implementation changes.

ezyang replied:
Or maybe we shouldn't match everything, just a subset of the "most important" op calls.

ezyang commented on torch/csrc/jit/runtime/register_c10_ops.cpp:
```diff
@@ -15,7 +15,6 @@ namespace {
 //      It should also handle autograd.
 Operator createOperatorFromC10(const c10::OperatorHandle& op) {
   return Operator(op, [op](Stack& stack) {
-    RECORD_FUNCTION(op.schema().name(), stack);
```
Interesting, what's going on here?

jamesr66a replied:
This is now covered by the dispatch wrapper when we do `callBoxed` below

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -317,6 +317,29 @@
 
+# ProfiledType templates
+PROFILE_DISPATCH_UNBOXED = CodeTemplate("""\
+auto op = c10::Dispatcher::singleton().findSchema({"aten::${operator_name}", "${overload_name}"});
```
This should probably be `static auto op` so you don't have to do a find schema every time you run this

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -545,16 +571,70 @@ def gen_variable_type_shard(out, aten_declarations, template_path, suffix, heade
                 wrapper_registrations.append(UNBOXEDONLY_WRAPPER_REGISTRATION.substitute(
                     declaration, formal_types=formal_types))
 
+        # TODO: ????? type mismatch in generated wrapper?
```
if you could paste the error here that would be helpful

jamesr66a replied:
Rebased and tried it again and it works now ¯\_(ツ)_/¯

ezyang commented on tools/autograd/templates/ProfiledType.cpp:
```diff
@@ -0,0 +1,50 @@
+// file is only for convenience; it's not actually used in the
+// build. If the file you're looking at now is one of the shards, you
+// may want to switch over to the Everything variant to make you
+// grepping smoother.
```
You can probably axe this comment and just reference to `Sharded File` in VariableType

ezyang commented:
This is looking really good. I would like to get some answers on the "big honking slow code" that has been copy pasted into (what looks like) a prod path, but this looks very good. I didn't review the codegen code too carefully at the moment.

Commit:
Update on "Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

jamesr66a commented:
After rebasing, now depends on https://github.com/pytorch/pytorch/pull/36398 for various functions to work in `callUnboxedRedispatch`

Commit:
Update on "Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

Commit:
Update on "Move profiler to a dispatch wrapper"


This makes it so that the profiler is not called from the VariableType wrappers, but rather from a boxed wrapper that calls RECORD_FUNCTION, then redispatches to the underlying operator.

Benchmark script:

```
import torch, time

x, y = torch.rand(1), torch.rand(1)

NITER = 10000

s = time.time()
for _ in range(NITER):
    x + y
e = time.time()

print('time per iter unprofiled', (e - s) / NITER)

s = time.time()
with torch.autograd.profiler.profile() as prof:
    for _ in range(NITER):
        x + y
e = time.time()
print('time per iter profiled', (e - s) / NITER)
```

Before this change
```
time per iter unprofiled 2.6654481887817384e-06
time per iter profiled 1.276531219482422e-05
```

After this change
```
time per iter unprofiled 2.346181869506836e-06
time per iter profiled 3.5948777198791505e-05
```

Differential Revision: [D19775659](https://our.internmc.facebook.com/intern/diff/D19775659)

[ghstack-poisoned]

kostmo commented:
Master breakage looks related to this PR:

```
[...]/ProfiledType_2.cpp:1754:26: error: ‘import’ is not a member of ‘torch’
```
